server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java

```diff
@@ -120,6 +149,10 @@ public NumberFieldMapper build(ContentPath contentPath) {
         }
     }
 
+    public interface MapperScript {
+        void executeAndEmit(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter);
```

Comment:
the emit terminology is already used for the emit function in scripts, and given that we have a consumer here, would it make sense to call this e.g. consumeValuesForDoc ?

Reply:
So I've moved this to a top-level class `FieldValues<T>` with a `valuesForDoc()` method.  It's a nice abstraction, actually, and I can see it being useful in other places as well later on.